### PR TITLE
Add sitewide listening activity statistics

### PR DESCRIPTION
**Pull request by amCap1712**
Add sitewide listening activity statistics

### CODE REVIEW

File: listenbrainz_spark/stats/common/listening_activity.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+import logging
+from datetime import datetime, time, date
+from typing import Tuple
+
+from dateutil.relativedelta import relativedelta, MO
```
Comment (mayhem) [inline listenbrainz_spark/stats/common/listening_activity.py:5]: MO is poorly named and even more poorly documented. Sigh.

File: listenbrainz_spark/stats/common/listening_activity.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        return relativedelta(years=-1, month=10, day=1)
+    elif month <= 9:
+        # currently, in Jul-Sep so 2 quarters ago is Jan-Mar
+        return relativedelta(month=1, day=1)
+    else:
+        # currently, in Oct-Dec so 2 quarters ago is Apr-Jun
+        return relativedelta(month=4, day=1)
+
+
+# other stats uses a different function (get_dates_for_stats_range) to calculate
```
Comment (mayhem) [inline listenbrainz_spark/stats/common/listening_activity.py:58]: move to docstring?

File: listenbrainz_spark/stats/common/listening_activity.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    to_date = from_date + to_offset
+
+    # set time to 00:00
+    from_date = datetime.combine(from_date, time.min)
+    to_date = datetime.combine(to_date, time.min)
+
+    return from_date, to_date, step, date_format
+
+
+def setup_time_range(stats_range: str) -> Tuple[datetime, datetime]:
```
Comment (mayhem) [inline listenbrainz_spark/stats/common/listening_activity.py:123]: docstring missing.

File: listenbrainz_spark/stats/common/listening_activity.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    from_date, to_date, step, date_format = get_time_range(stats_range)
+    time_range = []
+
+    period_start = from_date
```
Comment (mayhem) [inline listenbrainz_spark/stats/common/listening_activity.py:127]: I'm not a fan of period_start as the name here... segment_start?

File: listenbrainz_spark/stats/user/tests/test_listening_activity.py [hunk 5]
```diff
@@ -78,7 +79,7 @@ def test_get_listening_activity_year(self, mock_create_messages, _, mock_get_lis
         mock_create_messages.assert_called_with(data='activity_table', stats_range='year',
                                                 from_date=from_date, to_date=to_date)
 
-    @patch("listenbrainz_spark.stats.user.listening_activity.get_latest_listen_ts")
+    @patch("listenbrainz_spark.stats.common.listening_activity.get_latest_listen_ts")
```
Comment (mayhem) [inline listenbrainz_spark/stats/user/tests/test_listening_activity.py:82]: Should this test test the other time_ranges as well? I see tests for only quarter and half_yearly. These date function always scare me, so more tests could really be appropriate here.

File: listenbrainz_spark/stats/common/listening_activity.py (hunk 3 above)
Comment (mayhem) [reply listenbrainz_spark/stats/common/listening_activity.py:123]: This function is rather critical to how this PR works, so I think more docs are deffo in order.

File: listenbrainz_spark/stats/sitewide/listening_activity.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    """ Calculate number of listens for each user in time ranges given in the "time_range" table.
+    The time ranges are as follows:
+        1) week - each day with weekday name of the past 2 weeks.
+        2) month - each day the past 2 months.
+        3) year - each month of the past 2 years.
+        4) all_time - each year starting from LAST_FM_FOUNDING_YEAR (2002)
+    """
+    # calculates the number of listens in each time range for each user, count(listen.listened_at) so that
+    # group without listens are counted as 0, count(*) gives 1.
+    result = run_query("""
```
Comment (mayhem) [inline listenbrainz_spark/stats/sitewide/listening_activity.py:35]: I like how you're using the time_range table here -- clever!

**Review by mayhem — CHANGES_REQUESTED** [review]
Some more docs, please!

File: listenbrainz_spark/stats/user/tests/test_listening_activity.py (hunk 5 above)
Comment (amCap1712) [reply listenbrainz_spark/stats/user/tests/test_listening_activity.py:82]: Added tests for all ranges.

File: listenbrainz_spark/stats/sitewide/listening_activity.py (hunk 6 above)
Comment (amCap1712) [reply listenbrainz_spark/stats/sitewide/listening_activity.py:35]: 😄

**Review by mayhem — APPROVED** [review]
Thanks!

**Pull request merged**